### ISSUE TITLE
Phrase-boosting feature for Parakeet v2

### ISSUE THREAD

**Issue by hoavt-54** [NVIDIA/NeMo#14500]
Thanks @andrusenkoau for the recent phrase boosting features in ASR. I've tried using that for a while but only was able to boost phrases with AED decoding (Canary-1b). It does not really work for Parakeet v2 or I am missing something. WER gets worst as boosting_tree_alpha increases. 

```
os.environ["MODEL_NAME"] = "/home/jovyan/.cache/huggingface/hub/models--nvidia--parakeet-tdt-0.6b-v2/snapshots/f1cd6697a2ec38060af8a36ce206a4ad9c4467bc/parakeet-tdt-0.6b-v2.nemo"
os.environ["BATCH_SIZE"] = "1"
os.environ["BEAM_SIZE"] = "5"
os.environ["KEY_WORDS_LIST"] = cb_list_file
os.environ["HYDRA_FULL_ERROR"] = "1"
!python examples/asr/speech_to_text_eval.py \
        model_path=${MODEL_NAME} \
        dataset_manifest=./context_biasing_data/gtc_data_subset_10f.json \
        batch_size=${BATCH_SIZE} \
        output_filename=boosted_output.txt \
        text_processing.do_lowercase=true \
        text_processing.rm_punctuation=true \
        rnnt_decoding.strategy="malsd_batch" \ # greedy, greedy_batch
        # rnnt_decoding.beam.beam_size=${BEAM_SIZE} \
        rnnt_decoding.beam.boosting_tree.key_phrases_file=${KEY_WORDS_LIST} \
        rnnt_decoding.beam.boosting_tree.context_score=1.0 \
        rnnt_decoding.beam.boosting_tree.depth_scaling=2.0 \
        rnnt_decoding.beam.boosting_tree_alpha=0.6

```

**Phrase-boosting data test**: https://asr-tutorial-data.s3.eu-north-1.amazonaws.com/context_biasing_data.gz from [ASR_Context_Biasing.ipynb](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/ASR_Context_Biasing.ipynb). By the way if we could have a tutorial like the current ASR_Context_Biasing.ipynb, that would be great.

**boost_keywords.txt:**
```
gpu
nvidia
nvidia's
nvlink
omniverse
cunumeric
numpy
dgx
dgxs
dlss
cpu
tsmc
culitho
xlabs
tensorrt
tensorflow
pytorch
aws
chatgpt
pcie

```
**WER and CER results:**
```
boosting_tree_alpha     outputs
0.0                Dataset WER/CER 12.06%/4.96%
0.1                Dataset WER/CER 13.07%/5.05%
0.2                Dataset WER/CER 13.07%/5.05%
0.3                Dataset WER/CER 12.06%/4.96%
0.4                Dataset WER/CER 12.06%/4.96%
0.5                Dataset WER/CER 13.07%/6.03%
0.6                Dataset WER/CER 13.07%/6.03%
0.7                Dataset WER/CER 14.07%/6.12%
0.8                Dataset WER/CER 14.07%/6.12%
1.0                Dataset WER/CER 14.07%/6.38%
2.0                Dataset WER/CER 22.61%/14.98%
```

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- OS version: Linux jupyterserver-6f8d46c9c4-6s6nj 4.18.0-553.40.1.el8_10.x86_64
- PyTorch version: '2.7.1+cu126'
- Python version: 3.11
- GPU: L4
- branch: main

**Comment by Jacob-Bishop** [comment]
Ran into this as well. 

Don't have a reproducible example, but if I recall correctly, if you set boosting_tree_alpha very high (like >5) you start to see the boosted phrases show up where they aren't supposed to as you would expect for an over-boosted phrase, so the code is doing something...

**Comment by khoa-lak** [comment]
Do we have any documents for the parameters used in those phrase boosting functions? I can't get them to work.

**Comment by hoavt-54** [comment]
@nithinraok  When you get a chance, could you assign this to someone?

**Comment by andrusenkoau** [comment]
Hi @hoavt-54, I apologize for the delayed response.

Most likely, the main problem in your case is the `parakeet-tdt-0.6b-v2` model, which was trained to predict capitalized words. In this case, the context words `gpu` and `omniverse` can be recognized by the model as `GPU` and `Omniverse`. The proposed phrase boosting method changes the model probabilities at the token level; however, `gpu` and `GPU` will consist of different tokens. In this case, you need to add capitalized versions of all words (including abbreviations) to the context list. You can do this manually or with the help of an LLM.

To test this behavior further, you can try running the same pipeline with the original context list for https://huggingface.co/nvidia/parakeet-tdt-1.1b, which does not support word capitalization.

BTW, for `parakeet-tdt-0.6b-v2` model I would recommend using some other test set, since this model most likely recognizes data from https://asr-tutorial-data.s3.eu-north-1.amazonaws.com/context_biasing_data.gz with almost no errors by default.

**Comment by andrusenkoau** [comment]
Hi @khoa-lak, сan you describe your problem in more detail so that we can help you?

We plan to add the documentation later. For now, you can read the description of the proposed method in our paper - https://arxiv.org/abs/2508.07014.

**Comment by hoavt-54** [comment]
> Hi [@hoavt-54](https://github.com/hoavt-54), I apologize for the delayed response.
> 
> Most likely, the main problem in your case is the `parakeet-tdt-0.6b-v2` model, which was trained to predict capitalized words. In this case, the context words `gpu` and `omniverse` can be recognized by the model as `GPU` and `Omniverse`. The proposed phrase boosting method changes the model probabilities at the token level; however, `gpu` and `GPU` will consist of different tokens. In this case, you need to add capitalized versions of all words (including abbreviations) to the context list. You can do this manually or with the help of an LLM.
> 
> To test this behavior further, you can try running the same pipeline with the original context list for https://huggingface.co/nvidia/parakeet-tdt-1.1b, which does not support word capitalization.
> 
> BTW, for `parakeet-tdt-0.6b-v2` model I would recommend using some other test set, since this model most likely recognizes data from https://asr-tutorial-data.s3.eu-north-1.amazonaws.com/context_biasing_data.gz with almost no errors by default.

Thank you for your response @andrusenkoau, I've tried and the feature works pretty well for `parakeet-tdt-1.1b.` 

For `parakeet-tdt-0.6b-v2`, I find it pretty tricky to find the right capitalized versions of the word. The outputs are not always all-capitalized. For example:

```
CooNumeric accelerates NumPy
NVIDIA's TensorRT compiler
Nvidia Modulus, the PhysicsML framework and Omniverse
```
I'm wondering if including all possible capitalized versions to the context would make it worse than the right set capitalized versions?

**Comment by zhenyih** [comment]
ai generated, please verify

To properly configure phrase boosting for Parakeet-tdt-0.6b-v2, you need to include multiple capitalization variants of each keyword in your boosting list file.

Since this model was trained to predict capitalized words, the boosting mechanism works at the token level where "gpu" and "GPU" are treated as different tokens. This explains why your WER gets worse as boosting_tree_alpha increases - you're boosting the wrong token forms.

For your keywords list, include the most likely capitalization variants:

1. For standard words: include lowercase and title case (gpu, Gpu)
2. For acronyms: include both uppercase and spelled versions (GPU, G P U)
3. For brand names: include common variants (NVIDIA, Nvidia, nvidia)
4. For compound terms: include variants with different capitalization patterns (TensorRT, tensorrt, Tensorrt)

For example:
```
gpu
GPU
G P U
nvidia
Nvidia
NVIDIA
tensorrt
TensorRT
```

Start with a moderate boosting_tree_alpha value (0.3-0.5) and adjust based on results. Including too many variants might dilute the boosting effect, so prioritize the most common forms for your specific audio content.

For comparison, Parakeet-tdt-1.1b works with lowercase keywords because it doesn't support word capitalization.

**Comment by andrusenkoau** [comment]
> Thank you for your response [@andrusenkoau](https://github.com/andrusenkoau), I've tried and the feature works pretty well for `parakeet-tdt-1.1b.`
> 
> For `parakeet-tdt-0.6b-v2`, I find it pretty tricky to find the right capitalized versions of the word. The outputs are not always all-capitalized. For example:
> 
> ```
> CooNumeric accelerates NumPy
> NVIDIA's TensorRT compiler
> Nvidia Modulus, the PhysicsML framework and Omniverse
> ```
> 
> I'm wondering if including all possible capitalized versions to the context would make it worse than the right set capitalized versions?

This boosting method is quite robust to the growth of the number of words in the context list. I also tried adding capitalized keywords for `parakeet-tdt-0.6b-v2` in the case of the internal test set. This helped me improve the boosting results. But unfortunately, I don't know how to determine "the right set" automatically.

**Comment by andrusenkoau** [comment]
Hi @zhenyih!
This is a good answer for AI. However, it is essential to remember that adding the `G P U` phrase will not improve the probability of recognizing the word `GPU`. This phrase will increase the probability of the `G P U` transcription, since this boosting method only reweights the probabilities of tokens during decoding.
To use alternative BPE tokenizations for the same word, you can use the built-in parameter `use_bpe_dropout=True` in `BoostingTreeModelConfig`. It will generate several alternative tokenizations for your word.

**Comment by hoavt-54** [comment]
> > Thank you for your response [@andrusenkoau](https://github.com/andrusenkoau), I've tried and the feature works pretty well for `parakeet-tdt-1.1b.`
> > For `parakeet-tdt-0.6b-v2`, I find it pretty tricky to find the right capitalized versions of the word. The outputs are not always all-capitalized. For example:
> > ```
> > CooNumeric accelerates NumPy
> > NVIDIA's TensorRT compiler
> > Nvidia Modulus, the PhysicsML framework and Omniverse
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm wondering if including all possible capitalized versions to the context would make it worse than the right set capitalized versions?
> 
> This boosting method is quite robust to the growth of the number of words in the context list. I also tried adding capitalized keywords for `parakeet-tdt-0.6b-v2` in the case of the internal test set. This helped me improve the boosting results. But unfortunately, I don't know how to determine "the right set" automatically.

Thank @andrusenkoau, that works quite well when I add a large set of possible uppercase/lowercase of each word. 

Just one more thing tho, I see the model load and build the BoostingTreeModel every time transcribe() is called. I'm just worried this might slow thing down. This is how I tried:
```
            
            if self.boosting_keywords:
                asr_model.cfg.decoding.strategy = "greedy_batch"
                asr_model.cfg.decoding.greedy.boosting_tree.key_phrases_file=self.boosting_keywords
                asr_model.cfg.decoding.greedy.boosting_tree.context_score=CONTEXT_SCORE
                asr_model.cfg.decoding.greedy.boosting_tree.depth_scaling=DEPTH_SCALING
                asr_model.cfg.decoding.greedy.boosting_tree_alpha=BOOSTING_ALPHA

                asr_model.change_decoding_strategy(asr_model.cfg.decoding)
```

Even when I create and save the tree and set the model_path, it still has to read from disk.
 asr_model.cfg.decoding.greedy.boosting_tree.model_path

Do you know if we could just put in an in-memory model?

**Comment by andrusenkoau** [comment]
Hi @hoavt-54,
If you mean loading the tree into memory once and then running multiple decoding runs with the same tree, the current version of the script doesn't support that. To reduce tree building time, you can continue using the already built tree by specifying `boosting_tree.model_path`.

Have you noticed the difference in run time between decoding with and without a boosting tree? It seemed to me that tree building is very fast and doesn't noticeably compare to the decoding time.

**Comment by hoavt-54** [comment]
Hi @andrusenkoau 

>Have you noticed the difference in run time between decoding with and without a boosting tree? It seemed to me that tree building is very fast and doesn't noticeably compare to the decoding time.

that's true, building tree is really fast. Not much difference. I'll close this issue. Thank you guys for your support.
